File: packages/qcraft-engine/src/qcraft_engine/fiscal.py
"""Fiscal projection module (baseline_country).

Computes the recursive fiscal projection: revenue, expenditure, debt dynamics,
fiscal rule feedback, and derived indicators for years 2009-2099. This is the
most complex module in Q-CRAFT — every variable depends on prior-year state.

Uses explicit Python for-loops for the recursive computation (domain rule #1).
"""

import math

import polars as pl

YEAR_START = 2009
YEAR_END = 2099


def baseline_country(
    data_baseline: pl.DataFrame,
    data_interest: pl.DataFrame,
    data_macrofiscal: pl.DataFrame,
    debt_target: float,
    fiscal_rule: str,
    iso3c: str,
) -> pl.DataFrame:
    """Compute baseline fiscal projections for a single country.

    Args:
        data_baseline: Output of baseline_v1(). Must have columns:
            years, nominal_gdp, nominal_gdp_growth_percent,
            labour_productivity_growth, gdp_deflator_growth_percent,
            population_growth.
        data_interest: Output of interest_rate_country(). Must have columns:
            years, nominal_interest_rate.
        data_macrofiscal: Historical macrofiscal data (WEO period). Must have
            columns: iso3c, years, revenue, primary_expenditure, primary_balance,
            overall_balance, debt_to_gdp, debt, interest_expenditure, nominal_gdp,
            and corresponding _percent_gdp columns.
        debt_target: Debt-to-GDP target for fiscal rule (default 60).
        fiscal_rule: "Yes" or "No".
        iso3c: 3-letter ISO country code.

    Returns:
        DataFrame with 16 columns: years, revenue, revenue_percent_gdp,
        primary_expenditure, primary_expenditure_percent_gdp, primary_balance,
        primary_balance_percent_gdp, interest_expenditure,
        interest_expenditure_percent_gdp, total_expenditure, overall_balance,
        overall_balance_percent_gdp, debt_to_gdp, debt,
        debt_stabilizing_primary_balance, fiscal_gap.
        Years 2009-2099 (91 rows).
    """
    # Determine WEO_MAX_YEAR from macrofiscal data
    macro_country = data_macrofiscal.filter(pl.col("iso3c") == iso3c).sort("years")
    weo_max_year = int(macro_country["years"].max())  # type: ignore[arg-type]

    # Build lookups from macrofiscal (WEO period)
    macro_lookup: dict[int, dict[str, float]] = {}
    for row in macro_country.iter_rows(named=True):
        y = int(row["years"])
        macro_lookup[y] = {
            "revenue": float(row["revenue"]),
            "revenue_percent_gdp": float(row["revenue_percent_gdp"]),
            "primary_expenditure": float(row["primary_expenditure"]),
            "primary_expenditure_percent_gdp": float(
                row["primary_expenditure_percent_gdp"]
            ),
            "primary_balance": float(row["primary_balance"]),
            "primary_balance_percent_gdp": float(row["primary_balance_percent_gdp"]),
            "interest_expenditure": float(row["interest_expenditure"]),
            "interest_expenditure_percent_gdp": float(
                row["interest_expenditure_percent_gdp"]
            ),
            "total_expenditure": float(row["total_expenditure"]),
            "overall_balance": float(row["overall_balance"]),
            "overall_balance_percent_gdp": float(row["overall_balance_percent_gdp"]),
            "debt_to_gdp": float(row["debt_to_gdp"]),
            "debt": float(row["debt"]),
            "nominal_gdp": float(row["nominal_gdp"]),
        }

    # Build lookups from baseline_v1
    baseline = data_baseline.filter(
        pl.col("years").is_between(YEAR_START, YEAR_END)
    ).sort("years")
    nominal_gdp_lookup: dict[int, float] = {}
    nominal_gdp_growth_lookup: dict[int, float] = {}
    productivity_growth_lookup: dict[int, float] = {}
    inflation_lookup: dict[int, float] = {}
    pop_growth_lookup: dict[int, float] = {}
    for row in baseline.iter_rows(named=True):
        y = int(row["years"])
        nominal_gdp_lookup[y] = float(row["nominal_gdp"])
        nominal_gdp_growth_lookup[y] = float(row["nominal_gdp_growth_percent"])
        productivity_growth_lookup[y] = float(row["labour_productivity_growth"])
        inflation_lookup[y] = float(row["gdp_deflator_growth_percent"])
        pop_growth_lookup[y] = float(row["population_growth"])

    # Build lookup from interest rate
    interest = data_interest.filter(
        pl.col("years").is_between(YEAR_START, YEAR_END)
    ).sort("years")
    interest_rate_lookup: dict[int, float] = {}
    for row in interest.iter_rows(named=True):
        interest_rate_lookup[int(row["years"])] = float(row["nominal_interest_rate"])

    # Output arrays
    years_out = list(range(YEAR_START, YEAR_END + 1))
    n = len(years_out)

    revenue = [0.0] * n
    revenue_pct = [0.0] * n
    primary_exp = [0.0] * n
    primary_exp_pct = [0.0] * n
    primary_bal = [0.0] * n
    primary_bal_pct = [0.0] * n
    interest_exp = [0.0] * n
    interest_exp_pct = [0.0] * n
    total_exp = [0.0] * n
    overall_bal = [0.0] * n
    overall_bal_pct = [0.0] * n
    debt_to_gdp = [0.0] * n
    debt = [0.0] * n
    dspb: list[float | None] = [None] * n
    fiscal_gap: list[float | None] = [None] * n
    fiscal_rule_value = [0.0] * n

    # Fill WEO period from macrofiscal
    for i, year in enumerate(years_out):
        if year > weo_max_year:
            break

        m = macro_lookup[year]
        revenue[i] = m["revenue"]
        revenue_pct[i] = m["revenue_percent_gdp"]
        primary_exp[i] = m["primary_expenditure"]
        primary_exp_pct[i] = m["primary_expenditure_percent_gdp"]
        primary_bal[i] = m["primary_balance"]
        primary_bal_pct[i] = m["primary_balance_percent_gdp"]
        interest_exp[i] = m["interest_expenditure"]
        interest_exp_pct[i] = m["interest_expenditure_percent_gdp"]
        total_exp[i] = m["total_expenditure"]
        overall_bal[i] = m["overall_balance"]
        overall_bal_pct[i] = m["overall_balance_percent_gdp"]
        debt_to_gdp[i] = m["debt_to_gdp"]
        debt[i] = m["debt"]

        # DSPB: computed from 2010 onward (needs t-1 data)
        if i > 0:
            nom_rate = interest_rate_lookup[year]
            gdp_g = nominal_gdp_growth_lookup[year]
            dspb[i] = debt_to_gdp[i - 1] * (nom_rate - gdp_g) / 100 / (1 + gdp_g / 100)

        # Fiscal gap: needs DSPB
        if dspb[i] is not None:
            ngdp = nominal_gdp_lookup[year]
            dspb_val: float = dspb[i]  # type: ignore[assignment]
            fg = (primary_bal_pct[i] - dspb_val) / 100 * ngdp
            fiscal_gap[i] = fg

            # Fiscal rule during WEO period
            if fiscal_rule == "No":
                fiscal_rule_value[i] = 0.0
            elif i > 0:
                rising = debt_to_gdp[i] > debt_to_gdp[i - 1]
                above_target = debt_to_gdp[i] > debt_target
                if (rising and above_target) or (not rising and not above_target):
                    fiscal_rule_value[i] = fg
                else:
                    fiscal_rule_value[i] = 0.0

    # Recursive projection beyond WEO
    for i, year in enumerate(years_out):
        if year <= weo_max_year:
            continue

        ngdp = nominal_gdp_lookup[year]
        ngdp_growth = nominal_gdp_growth_lookup[year]
        prod_growth = productivity_growth_lookup[year]
        infl = inflation_lookup[year]
        pop_growth = pop_growth_lookup[year]
        nom_rate = interest_rate_lookup[year]

        # Step 1: Revenue
        revenue[i] = revenue[i - 1] * (1 + ngdp_growth / 100)
        revenue_pct[i] = revenue[i] / ngdp * 100

        # Step 2: Primary expenditure (multiplicative growth + fiscal rule lag)
        primary_exp[i] = (
            primary_exp[i - 1]
            * (1 + prod_growth / 100)
            * (1 + infl / 100)
            * (1 + pop_growth / 100)
            + fiscal_rule_value[i - 1]
        )
        primary_exp_pct[i] = primary_exp[i] / ngdp * 100

        # Step 3: Primary balance
        primary_bal[i] = revenue[i] - primary_exp[i]
        primary_bal_pct[i] = primary_bal[i] / ngdp * 100

        # Step 4: Debt-to-GDP (baseline applies max(0, ...))
        raw_debt_to_gdp = (
            debt_to_gdp[i - 1] * (1 + nom_rate / 100) / (1 + ngdp_growth / 100)
            - primary_bal_pct[i]
        )
        debt_to_gdp[i] = max(0.0, raw_debt_to_gdp)

        # Step 5: Debt level
        debt[i] = debt_to_gdp[i] / 100 * ngdp

        # Step 6: Interest expenditure (prior-year debt * current rate)
        interest_exp[i] = debt[i - 1] * nom_rate / 100
        interest_exp_pct[i] = interest_exp[i] / ngdp * 100

        # Step 7: Total expenditure and overall balance
        total_exp[i] = primary_exp[i] + interest_exp[i]
        overall_bal[i] = revenue[i] - total_exp[i]
        overall_bal_pct[i] = overall_bal[i] / ngdp * 100

        # Step 8: DSPB
        dspb[i] = (
            debt_to_gdp[i - 1]
            * (nom_rate - ngdp_growth)
            / 100
            / (1 + ngdp_growth / 100)
        )

        # Step 9: Fiscal gap
        fg = (primary_bal_pct[i] - dspb[i]) / 100 * ngdp  # type: ignore[operator]
        fiscal_gap[i] = fg

        # Step 10: Fiscal rule value (applied to NEXT year's expenditure)
        if fiscal_rule == "No":
            fiscal_rule_value[i] = 0.0
        else:
            rising = debt_to_gdp[i] > debt_to_gdp[i - 1]
            above_target = debt_to_gdp[i] > debt_target
            if (rising and above_target) or (not rising and not above_target):
                fiscal_rule_value[i] = fg
            else:
                fiscal_rule_value[i] = 0.0

    # Convert NaN fiscal_gap values to None for null representation
    fiscal_gap_clean: list[float | None] = [
        None if v is None or (isinstance(v, float) and math.isnan(v)) else v
        for v in fiscal_gap
    ]

    return pl.DataFrame(
        {
            "years": years_out,
            "revenue": revenue,
            "revenue_percent_gdp": revenue_pct,
            "primary_expenditure": primary_exp,
            "primary_expenditure_percent_gdp": primary_exp_pct,
            "primary_balance": primary_bal,
            "primary_balance_percent_gdp": primary_bal_pct,
            "interest_expenditure": interest_exp,
            "interest_expenditure_percent_gdp": interest_exp_pct,
            "total_expenditure": total_exp,
            "overall_balance": overall_bal,
            "overall_balance_percent_gdp": overall_bal_pct,
            "debt_to_gdp": debt_to_gdp,
            "debt": debt,
            "debt_stabilizing_primary_balance": dspb,
            "fiscal_gap": fiscal_gap_clean,
        }
    )

File: packages/qcraft-engine-ts/src/fiscal.ts
/**
 * Fiscal projection module (`baseline_country` in Python).
 *
 * Computes the recursive fiscal projection: revenue, expenditure, debt dynamics,
 * fiscal rule feedback, and derived indicators for 2009–2099. Every projection-period
 * variable depends on prior-year state, so this is an explicit year-by-year loop
 * (domain rule #1) — never a vectorised shift/cumsum.
 */

import type {
  BaselineV1Row,
  FiscalRow,
  FiscalRuleSetting,
  InterestRateRow,
  MacroFiscalRow,
  Num,
} from './types.js';
import { YEAR_END, YEAR_START } from './constants.js';
import { mustGet, nanToNull } from './internal.js';
import { MissingDebtAnchorError, MissingMacrofiscalInputError } from './errors.js';

/**
 * The macrofiscal columns this module reads. The row it is handed carries more
 * than these (`interest_rate_percent` among them, which the interest rate module
 * consumes and this one does not), so a null is only this module's problem when
 * it lands in a column this module reads.
 */
const READS = [
  'revenue',
  'revenue_percent_gdp',
  'primary_expenditure',
  'primary_expenditure_percent_gdp',
  'primary_balance',
  'primary_balance_percent_gdp',
  'interest_expenditure',
  'interest_expenditure_percent_gdp',
  'total_expenditure',
  'overall_balance',
  'overall_balance_percent_gdp',
  'debt_to_gdp',
  'debt',
  'nominal_gdp',
] as const satisfies readonly (keyof MacroFiscalRow)[];

/** `null` coerces to 0 in JavaScript arithmetic, so it has to be caught by value. */
function isNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value);
}

export interface FiscalOptions {
  /** Debt-to-GDP target for the fiscal rule. */
  debtTarget: number;
  /** "Yes" enables the fiscal rule feedback loop; "No" zeroes it. */
  fiscalRule: FiscalRuleSetting;
}

/**
 * Compute baseline fiscal projections for a single country.
 *
 * @param dataBaseline Output of `baselineV1`.
 * @param dataInterest Output of `interestRateCountry`.
 * @param dataMacrofiscal Historical (WEO-period) fiscal data.
 * @returns 91 rows (2009–2099) with 16 columns.
 */
export function baselineCountry(
  dataBaseline: readonly BaselineV1Row[],
  dataInterest: readonly InterestRateRow[],
  dataMacrofiscal: readonly MacroFiscalRow[],
  iso3c: string,
  options: FiscalOptions,
): FiscalRow[] {
  const { debtTarget, fiscalRule } = options;

  const macroCountry = dataMacrofiscal
    .filter((r) => r.iso3c === iso3c)
    .sort((a, b) => a.years - b.years);
  if (macroCountry.length === 0) {
    throw new Error(`No data found for iso3c='${iso3c}' in data_macrofiscal`);
  }
  const weoMaxYear = Math.max(...macroCountry.map((r) => r.years));

  // The projection carries debt forward from the last WEO year, so that year's
  // debt ratio is the one number without which nothing downstream is defined.
  // The workbook does not guard this either: `Baseline` row 36 anchors on its
  // last WEO column and reads `#VALUE!` all the way to 2099 when the column is
  // empty. Python raises here; without this check JavaScript coerced the null to
  // zero and drew a debt path pinned at the floor.
  const anchorRow = macroCountry.find((r) => r.years === weoMaxYear);
  if (anchorRow === undefined || !isNumber(anchorRow.debt_to_gdp)) {
    throw new MissingDebtAnchorError(iso3c, weoMaxYear);
  }

  // Only the years the projection actually reads. The fill loop below starts at
  // YEAR_START and stops at the last WEO year, and the workbook's own Baseline
  // sheet starts its axis at 2009 too, so a hole in 2001-2008 is not this
  // engine's business.
  const macroLookup = new Map<number, MacroFiscalRow>();
  for (const row of macroCountry) {
    if (row.years < YEAR_START || row.years > weoMaxYear) continue;
    for (const field of READS) {
      if (!isNumber(row[field])) {
        throw new MissingMacrofiscalInputError(iso3c, row.years, field);
      }
    }
    macroLookup.set(row.years, row);
  }

  const nominalGdpLookup = new Map<number, number>();
  const nominalGdpGrowthLookup = new Map<number, number>();
  const productivityGrowthLookup = new Map<number, number>();
  const inflationLookup = new Map<number, number>();
  const popGrowthLookup = new Map<number, number>();
  for (const row of dataBaseline) {
    if (row.years < YEAR_START || row.years > YEAR_END) continue;
    nominalGdpLookup.set(row.years, row.nominal_gdp);
    nominalGdpGrowthLookup.set(row.years, row.nominal_gdp_growth_percent);
    productivityGrowthLookup.set(row.years, row.labour_productivity_growth);
    inflationLookup.set(row.years, row.gdp_deflator_growth_percent);
    popGrowthLookup.set(row.years, row.population_growth);
  }

  const interestRateLookup = new Map<number, number>();
  for (const row of dataInterest) {
    if (row.years >= YEAR_START && row.years <= YEAR_END) {
      interestRateLookup.set(row.years, row.nominal_interest_rate);
    }
  }

  const years: number[] = [];
  for (let y = YEAR_START; y <= YEAR_END; y += 1) years.push(y);
  const n = years.length;

  const revenue = new Array<number>(n).fill(0);
  const revenuePct = new Array<number>(n).fill(0);
  const primaryExp = new Array<number>(n).fill(0);
  const primaryExpPct = new Array<number>(n).fill(0);
  const primaryBal = new Array<number>(n).fill(0);
  const primaryBalPct = new Array<number>(n).fill(0);
  const interestExp = new Array<number>(n).fill(0);
  const interestExpPct = new Array<number>(n).fill(0);
  const totalExp = new Array<number>(n).fill(0);
  const overallBal = new Array<number>(n).fill(0);
  const overallBalPct = new Array<number>(n).fill(0);
  const debtToGdp = new Array<number>(n).fill(0);
  const debt = new Array<number>(n).fill(0);
  const dspb = new Array<Num>(n).fill(null);
  const fiscalGap = new Array<Num>(n).fill(null);
  /** Applied to the FOLLOWING year's primary expenditure — additive in levels, not a rate. */
  const fiscalRuleValue = new Array<number>(n).fill(0);

  // --- WEO period: take fiscal aggregates straight from macrofiscal ---
  for (let i = 0; i < n; i += 1) {
    const year = years[i]!;
    if (year > weoMaxYear) break;

    const m = mustGet(macroLookup, year, 'macrofiscal');
    revenue[i] = m.revenue;
    revenuePct[i] = m.revenue_percent_gdp;
    primaryExp[i] = m.primary_expenditure;
    primaryExpPct[i] = m.primary_expenditure_percent_gdp;
    primaryBal[i] = m.primary_balance;
    primaryBalPct[i] = m.primary_balance_percent_gdp;
    interestExp[i] = m.interest_expenditure;
    interestExpPct[i] = m.interest_expenditure_percent_gdp;
    totalExp[i] = m.total_expenditure;
    overallBal[i] = m.overall_balance;
    overallBalPct[i] = m.overall_balance_percent_gdp;
    debtToGdp[i] = m.debt_to_gdp;
    debt[i] = m.debt;

    // DSPB: computed from 2010 onward (needs t-1 debt).
    if (i > 0) {
      const nomRate = mustGet(interestRateLookup, year, 'nominal_interest_rate');
      const gdpG = mustGet(nominalGdpGrowthLookup, year, 'nominal GDP growth');
      dspb[i] = ((debtToGdp[i - 1]! * (nomRate - gdpG)) / 100) / (1 + gdpG / 100);
    }

    const dspbVal = dspb[i]!;
    if (dspbVal !== null) {
      const ngdp = mustGet(nominalGdpLookup, year, 'nominal_gdp');
      const fg = ((primaryBalPct[i]! - dspbVal) / 100) * ngdp;

      // Fiscal gap is only reported from WEO_MAX_YEAR - 3 onward.
      if (year >= weoMaxYear - 3) fiscalGap[i] = fg;

      if (fiscalRule === 'No') {
        fiscalRuleValue[i] = 0.0;
      } else if (i > 0) {
        const rising = debtToGdp[i]! > debtToGdp[i - 1]!;
        const aboveTarget = debtToGdp[i]! > debtTarget;
        fiscalRuleValue[i] =
          (rising && aboveTarget) || (!rising && !aboveTarget) ? fg : 0.0;
      }
    }
  }

  // --- Recursive projection beyond WEO (explicit t-1 loop, domain rule #1) ---
  for (let i = 0; i < n; i += 1) {
    const year = years[i]!;
    if (year <= weoMaxYear) continue;

    const ngdp = mustGet(nominalGdpLookup, year, 'nominal_gdp');
    const ngdpGrowth = mustGet(nominalGdpGrowthLookup, year, 'nominal GDP growth');
    const prodGrowth = mustGet(productivityGrowthLookup, year, 'productivity growth');
    const infl = mustGet(inflationLookup, year, 'inflation');
    const popGrowth = mustGet(popGrowthLookup, year, 'population growth');
    const nomRate = mustGet(interestRateLookup, year, 'nominal_interest_rate');

    // Step 1: Revenue grows with nominal GDP.
    revenue[i] = revenue[i - 1]! * (1 + ngdpGrowth / 100);
    revenuePct[i] = (revenue[i]! / ngdp) * 100;

    // Step 2: Primary expenditure — MULTIPLICATIVE growth (1+a)(1+b)(1+c), then the
    // prior year's fiscal rule adjustment added in LEVELS (domain rule #2).
    primaryExp[i] =
      primaryExp[i - 1]! *
        (1 + prodGrowth / 100) *
        (1 + infl / 100) *
        (1 + popGrowth / 100) +
      fiscalRuleValue[i - 1]!;
    primaryExpPct[i] = (primaryExp[i]! / ngdp) * 100;

    // Step 3: Primary balance.
    primaryBal[i] = revenue[i]! - primaryExp[i]!;
    primaryBalPct[i] = (primaryBal[i]! / ngdp) * 100;

    // Step 4: Debt-to-GDP — the BASELINE applies the max(0, ...) floor (domain rule #3).
    const rawDebtToGdp =
      (debtToGdp[i - 1]! * (1 + nomRate / 100)) / (1 + ngdpGrowth / 100) -
      primaryBalPct[i]!;
    debtToGdp[i] = Math.max(0.0, rawDebtToGdp);

    // Step 5: Debt level.
    debt[i] = (debtToGdp[i]! / 100) * ngdp;

    // Step 6: Interest expenditure on prior-year debt at the current rate.
    interestExp[i] = (debt[i - 1]! * nomRate) / 100;
    interestExpPct[i] = (interestExp[i]! / ngdp) * 100;

    // Step 7: Total expenditure and overall balance.
    totalExp[i] = primaryExp[i]! + interestExp[i]!;
    overallBal[i] = revenue[i]! - totalExp[i]!;
    overallBalPct[i] = (overallBal[i]! / ngdp) * 100;

    // Step 8: Debt-stabilising primary balance.
    const dspbVal = ((debtToGdp[i - 1]! * (nomRate - ngdpGrowth)) / 100) /
      (1 + ngdpGrowth / 100);
    dspb[i] = dspbVal;

    // Step 9: Fiscal gap.
    const fg = ((primaryBalPct[i]! - dspbVal) / 100) * ngdp;
    fiscalGap[i] = fg;

    // Step 10: Fiscal rule value, applied to NEXT year's expenditure.
    if (fiscalRule === 'No') {
      fiscalRuleValue[i] = 0.0;
    } else {
      const rising = debtToGdp[i]! > debtToGdp[i - 1]!;
      const aboveTarget = debtToGdp[i]! > debtTarget;
      fiscalRuleValue[i] =
        (rising && aboveTarget) || (!rising && !aboveTarget) ? fg : 0.0;
    }
  }

  return years.map((year, i) => ({
    years: year,
    revenue: revenue[i]!,
    revenue_percent_gdp: revenuePct[i]!,
    primary_expenditure: primaryExp[i]!,
    primary_expenditure_percent_gdp: primaryExpPct[i]!,
    primary_balance: primaryBal[i]!,
    primary_balance_percent_gdp: primaryBalPct[i]!,
    interest_expenditure: interestExp[i]!,
    interest_expenditure_percent_gdp: interestExpPct[i]!,
    total_expenditure: totalExp[i]!,
    overall_balance: overallBal[i]!,
    overall_balance_percent_gdp: overallBalPct[i]!,
    debt_to_gdp: debtToGdp[i]!,
    debt: debt[i]!,
    debt_stabilizing_primary_balance: dspb[i]!,
    fiscal_gap: fiscalGap[i] === null ? null : nanToNull(fiscalGap[i] as number),
  }));
}
